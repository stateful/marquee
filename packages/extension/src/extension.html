<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <meta
      http-equiv="Content-Security-Policy"
      content="
        default-src 'unsafe-inline';
        img-src 'self' data: <%= it.cspSource %> https:;
        font-src <%= it.cspSource %> https:;
        script-src 'nonce-<%= it.nonce %>';
        connect-src <%= it.connectSrc.join(' ') %> 'unsafe-eval';
        child-src <%= it.childSrc.join(' ') %>;
      "
    >
    <base href="<%= it.baseAppUri %>">
    <title>Marquee</title>
    <style>
        .MuiTypography-h1,
        .MuiTypography-h2,
        .MuiTypography-h3,
        .MuiTypography-h4,
        .MuiTypography-h5,
        .MuiTypography-h6,
        .MuiTypography-subtitle1,
        .MuiTypography-subtitle2,
        .MuiTypography-body1,
        .MuiTypography-body2,
        .MuiTypography-button,
        .MuiTypography-caption,
        .MuiTypography-overline {
            font-size: <%= it.fontSize %>em!important;
        }
    </style>
    <style>
      :root {
        --marquee-theme-color: <%= it.colorScheme %>;
      }
    </style>
    <script nonce="<%= it.nonce %>">
      function b64DecodeUnicode(str) {
        // Going backwards: from bytestream, to percent-encoding, to original string.
        return decodeURIComponent(atob(str).split('').map(function(c) {
            return '%' + ('00' + c.charCodeAt(0).toString(16)).slice(-2);
        }).join(''));
      }
      window.marqueeThirdPartyWidgets = <%= it.customWidgetCounter %>;
      window.activeWorkspace = <%~ JSON.stringify(it.aws) %>;
      window.marqueeUserProps = '<%~ JSON.stringify(it.props) %>';
      window.marqueeBackendBaseUrl = '<%= it.baseUrl %>';
      window.marqueeBackendGeoUrl = '<%= it.geoUrl %>';
      window.marqueeBackendFwdGeoUrl = '<%= it.fwdGeoUrl %>';
      window.marqueeStateConfiguration = JSON.parse(b64DecodeUnicode('<%= it.widgetStateConfigurations %>'));
    </script>
</head>
<body>
    <div id="root"></div>
    <script nonce="<%= it.nonce %>" src="<%= it.baseAppUri %>/main.js"></script>
    <% for (const widgetScript of it.widgetScripts) { %>
      <%~ widgetScript %>
    <% } %>
</body>
</html>
