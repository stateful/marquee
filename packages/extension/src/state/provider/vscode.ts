import type { ExtensionContext } from 'vscode'
import type { StateProvider } from '../types'

export class VSCodeState implements StateProvider {
  constructor (public context: ExtensionContext) {}

  getState (key: string) {
    return this.context.globalState.get(key) as never
  }

  async setState (key: string, value: any) {
    await this.context.globalState.update(key, value)
  }
}
