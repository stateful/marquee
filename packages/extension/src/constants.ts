import vscode from 'vscode';

export const DEFAULT_FONT_SIZE = 5;

export const RETRY = 4 * 1000;
export const INIT = 50;
export const config = vscode.workspace.getConfiguration('marquee');
export const FILE_FILTER = { 'Marquee Settings': ['json'] };
export const CONFIG_FILE_TYPE = 'MarqueeSettings';
export const THIRD_PARTY_EXTENSION_DIR = '3rdParty';
