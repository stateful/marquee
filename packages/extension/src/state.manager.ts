import vscode from "vscode";
import crypto from "crypto";
import { Subject, from } from "rxjs";
import {
  pluck,
  filter,
  map,
  catchError,
  throttleTime,
  publishBehavior,
  refCount,
} from "rxjs/operators";
import { Item } from "./tree.view";

export const persistenceKey = "persistence";

export interface Message {
  feedback?: object;
  fwdgeo?: object;
  activeWorkspaceId?: string | null;
  autoDetect?: boolean;
  execCommands?: Array<{ command: string; args: Array<object>, options?: any }>;
  globalScope?: boolean;
  geo?: object;
  language?: any;
  liked?: Array<string>;
  modeName?: string;
  modes?: Array<{ layouts: Array<object>; widgets: Array<object> }>;
  news?: Array<object>;
  snippets?: Array<object>;
  dialogs?: Array<object>;
  read?: Array<string>;
  since?: any;
  theme?: object;
  persistence?: object;
  ready?: boolean;
  scope?: boolean;
  todos?: Array<object>;
  notes?: Array<object>;
  tricks?: Array<object>;
  upvote?: object;
  notify?: { type: NotificationType; message: string };
  version?: string;
  east?: Message;
  west?: Message;
  widgets?: any;
  workspaces?: Array<object>;
}

export enum NotificationType {
  ERROR = "error",
}

export enum WorkspaceType {
  WORKSPACE = "workspace",
  FOLDER = "folder",
  NONE = "none",
}

export interface Workspace {
  id: string;
  name: string;
  type: WorkspaceType;
  path: string;
}

export interface Todo {
  archived: boolean;
  body: string;
  checked: boolean;
  id: string;
  path: string;
  exists?: boolean;
  language: string;
  workspaceId: string | undefined;
}

export interface Snippet {
  archived: boolean;
  title: string;
  body: string;
  createdAt: number;
  id: string;
  origin?: string;
  path: string;
  exists?: boolean;
  language: { name: string; value: string };
  workspaceId: string | undefined;
}

export type Note = Snippet;

export class StateManager {
  protected readonly internal = new Subject<Message>();
  public readonly state = this.internal.pipe(publishBehavior({}), refCount());

  constructor(private readonly context: vscode.ExtensionContext) {
    this.handlePersistence();
  }

  receive(msg: Message) {
    this.emit(msg);
  }

  emit(msg: Message) {
    this.internal.next(msg);
  }

  update(msg: Message) {
    const obj: Message = { ...msg };
    delete obj.persistence;
    const east: Message = { east: obj };
    this.emit(east);
  }

  handlePersistence() {
    const persistence$ = this.state.pipe(
      pluck(persistenceKey),
      filter((save: any) => save !== undefined)
    );
    persistence$
      .pipe(
        map((save) => {
          const obj = {
            ...this.recover(),
            ...save,
          };
          return obj;
        }),
        map((obj) => {
          return from(this.context.globalState.update(persistenceKey, obj));
        }),
        catchError((err: any) => {
          console.log(err);
          return err;
        })
      )
      .subscribe();
  }

  recover(): any {
    return this.context.globalState.get<object>(persistenceKey, {});
  }

  save(handler: Function) {
    const latest = this.recover();
    delete latest.persistence;

    const obj = handler(latest);

    from(this.context.globalState.update(persistenceKey, obj)).subscribe(() => {
      this.emit({ east: obj });
    });
  }

  handleUpdates(obs?: any) {
    const updates = this.state.pipe(
      map((msg) => {
        // console.log(new Date());
        // console.log(JSON.stringify(Object.keys(msg)));

        const obj: Message = { ...msg };
        delete obj.persistence;
        delete obj.ready;
        return obj;
      }),
      filter((obj) => {
        return Object.keys(obj).length > 0;
      }),
      throttleTime(50) // unlikely humans are faster
    );

    if (obs) {
      updates.subscribe(obs);
    }

    return updates;
    // .pipe(
    //   tap((obj) => {
    //     console.log(new Date());
    //     console.log(Object.keys(obj));
    //   })
    // );
  }

  handleItem(op: Function) {
    return (item?: Item) => {
      this.save((latest: any) => {
        const t = `${item?.type.toLowerCase()}s`;
        const target = latest[t];
        if (target && item) {
          const newarr = target.map(op(item));
          const ret = { ...latest };
          ret[t] = newarr;
          return ret;
        }
        return latest;
      });
    };
  }

  /**
   * ToDo(Christian): remove once storage migration finished
   */
  public updateWorkspaces(latest: Message): Message {
    if (latest.workspaces === undefined) {
      latest.workspaces = [];
    }

    const aws = this.getActiveWorkspace();
    if (
      latest.activeWorkspaceId !== null &&
      latest.activeWorkspaceId !== undefined
    ) {
      latest.activeWorkspaceId = undefined;
    }

    if (
      /**
       * the current workspace can be detected
       */
      aws &&
      /**
       * the workspace path is not a number
       * (happens e.g. when connecting to docker container)
       */
      isNaN(Number(aws.path)) &&
      /**
       * the workspace isn't part of the existing list
       */
      !latest.workspaces.find((ws: any) => ws.id === aws.id) &&
      /**
       * we are not running on a remote machine, this is necessary
       * because we aren't able to connect to remote VSCode instances
       * through `vscode.openFolder`
       * see more: https://github.com/microsoft/vscode-remote-release/issues/6243
       */
      typeof vscode.env.remoteName === 'undefined'
    ) {
      latest.workspaces.push(aws);
    }

    return latest;
  }

  /**
   * ToDo(Christian): remove once storage migration finished
   */
  public getActiveWorkspace(): Workspace | null {
    const wsp = vscode.workspace;
    let name = wsp.name || "";
    let path = "";
    let type = WorkspaceType.NONE;

    if (wsp.workspaceFile) {
      type = WorkspaceType.WORKSPACE;
      path = wsp.workspaceFile.path;
    } else if (wsp.workspaceFolders) {
      type = WorkspaceType.FOLDER;
      path =
        wsp.workspaceFolders.length > 0 ? wsp.workspaceFolders[0].uri.path : "";
    }

    if (type && path) {
      const shasum = crypto.createHash("sha1");
      const id = shasum.update(path, "utf8").digest("hex");
      const nws: Workspace = { id, name, type, path };

      return nws;
    }

    return null;
  }
}
