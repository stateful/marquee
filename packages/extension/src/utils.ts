import vscode from 'vscode';
import ExtensionManager, { defaultConfigurations } from '@vscode-marquee/utils/extension';

import { MODES_UPDATE_TIMEOUT } from './constants';

export const isExpanded = (id: number): vscode.TreeItemCollapsibleState => {
  const found = [0, 1, 2, 3].filter((item: number) => id === item).length > 0;

  if (found) {
    return vscode.TreeItemCollapsibleState.Expanded;
  }

  return vscode.TreeItemCollapsibleState.Collapsed;
};

export const filterByScope = <T extends { workspaceId: string | null }>(
  obj: T[],
  aws: null | { id: string },
  globalScope: boolean
) => {
  return obj.filter((n) => globalScope || (aws && aws.id === n.workspaceId));
};

export const DEFAULT_STATE = {
  modeName: 'default',
  prevMode: null
};

export const DEFAULT_CONFIGURATION = {
  modes: defaultConfigurations['marquee.configuration.modes'].default,
  proxy: defaultConfigurations['marquee.configuration.proxy'].default,
  fontSize: defaultConfigurations['marquee.configuration.fontSize'].default,
  launchOnStartup: defaultConfigurations['marquee.configuration.launchOnStartup'].default,
  workspaceLaunch: defaultConfigurations['marquee.configuration.workspaceLaunch'].default,
  colorScheme: undefined
};

type Configuration = typeof DEFAULT_CONFIGURATION;
type State = typeof DEFAULT_STATE;

class GUIExtensionManager extends ExtensionManager<State, Configuration> {
  private _lastModesChange = Date.now();

  constructor (
    context: vscode.ExtensionContext,
    channel: vscode.OutputChannel,
    key: string,
    defaultConfiguration: Configuration,
    defaultState: State
  ) {
    super(context, channel, key, defaultConfiguration, defaultState);
    this._disposables.push(vscode.workspace.onDidChangeConfiguration(this._onModeChange.bind(this)));
  }

  async updateConfiguration <T extends keyof Configuration = keyof Configuration>(prop: T, val: Configuration[T]) {
    /**
     * when updating modes the webview passes the native ascii icon (e.g. "ðŸ’¼") which
     * breaks when sending from webview to extension host. This ensures that we don't
     * includes these broken ascii characters into the VSCode settings file
     */
    if (prop === 'modes') {
      this._lastModesChange = Date.now();

      for (const modeName of Object.keys(val)) {
        if (val[modeName].icon) {
          delete val[modeName].icon.native;
        }
      }
    }

    super.updateConfiguration(prop, val);
  }

  _onModeChange (event: vscode.ConfigurationChangeEvent) {
    if (
      !event.affectsConfiguration(`marquee.configuration.modes`) ||
      (Date.now() - this._lastModesChange) < MODES_UPDATE_TIMEOUT
    ) {
      return;
    }

    const config = vscode.workspace.getConfiguration('marquee');
    const val = config.get('configuration.modes') as Configuration[keyof Configuration];
    this._channel.appendLine(`Update configuration.modes via configuration listener`);
    this.broadcast({ modes: JSON.parse(JSON.stringify(val)) });
  }
}

export function activateGUI (
  context: vscode.ExtensionContext,
  channel: vscode.OutputChannel
) {
  const stateManager = new GUIExtensionManager(context, channel, 'configuration', DEFAULT_CONFIGURATION, DEFAULT_STATE);

  return {
    marquee: {
      disposable: stateManager,
      defaultState: stateManager.state,
      defaultConfiguration: stateManager.configuration,
      setup: stateManager.setBroadcaster.bind(stateManager)
    }
  };
}

export const linkMarquee = async (item: any) => {
  let file = item?.item?.origin;

  if (!file) {
    return;
  }

  let splt = file.split(":");
  let ln = "1";
  if (splt.length > 2) {
    ln = splt[splt.length - 1];
    splt = splt.splice(0, splt.length - 1);
    file = splt.join(":");
  } else if (splt.length > 1) {
    [file, ln] = splt;
  }
  const rpath = vscode.Uri.parse(file).fsPath;
  const doc = await vscode.workspace.openTextDocument(rpath);
  if (!doc) {
    return;
  }

  const editor = await vscode.window.showTextDocument(doc);
  const r = doc.lineAt(parseInt(ln)).range;
  if (!editor || !r) {
    return;
  }

  editor.revealRange(r, vscode.TextEditorRevealType.InCenter);
};
