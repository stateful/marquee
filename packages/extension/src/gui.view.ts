import fs from 'fs/promises'
import path from 'path'
import vscode from 'vscode'
import Channel from 'tangle/webviews'
import { v4 as uuidv4 } from 'uuid'
import { EventEmitter } from 'events'
import { render } from 'eta'
import { getExtProps } from '@vscode-marquee/utils/extension'
import type { Client } from 'tangle'
import type { MarqueeEvents } from '@vscode-marquee/utils'

import telemetry from './telemetry'
import StateManager from './stateManager'
import { DEFAULT_STATE } from './utils'
import { DEFAULT_FONT_SIZE, THIRD_PARTY_EXTENSION_DIR } from './constants'
import type { ExtensionConfiguration, ExtensionExport } from './types'

declare const BACKEND_BASE_URL: string
declare const BACKEND_GEO_URL: string
declare const BACKEND_FWDGEO_URL: string

export class MarqueeGui extends EventEmitter {
  private panel: vscode.WebviewPanel | null = null
  private guiActive: boolean = false
  private client?: Client<MarqueeEvents>

  private _baseUri = vscode.Uri.joinPath(this.context.extensionUri)
  private _template: Thenable<Uint8Array>
  private _templateDecoded?: string

  constructor (
    private readonly context: vscode.ExtensionContext,
    private readonly stateMgr: StateManager,
    private readonly channel: vscode.OutputChannel
  ) {
    super()
    this._template = vscode.workspace.fs.readFile(vscode.Uri.joinPath(this._baseUri, 'dist', 'extension.html'))

    /**
     * register listeners for Marquee state managers so they can interact with the gui
     */
    for (const widgetExtension of this.stateMgr.widgetExtensions) {
      widgetExtension.exports.marquee?.disposable?.on('gui.open', this.open.bind(this))
      widgetExtension.exports.marquee?.disposable?.on('gui.close', this.close.bind(this))
    }
  }

  async getTemplate () {
    if (this._templateDecoded) {
      return this._templateDecoded
    }

    const dec = new TextDecoder('utf-8')
    const tpl = await this._template
    this._templateDecoded = dec.decode(tpl)
    return this._templateDecoded
  }

  public isActive () {
    return this.guiActive
  }

  public close () {
    this.panel?.dispose()
  }

  public broadcast<EventName extends keyof MarqueeEvents>(
    event: EventName,
    payload: MarqueeEvents[EventName]
  ) {
    if (!this.client) {
      return false
    }
    this.client.emit(event, payload)
  }

  private async _verifyWidgetStates () {
    /**
     * check if current selected mode was removed and switch to the default
     * mode if so
     */
    if (this.stateMgr.gui.state.modeName && !this.stateMgr.gui.configuration.modes[this.stateMgr.gui.state.modeName]) {
      this.channel.appendLine(`Couldn't find selected mode "${this.stateMgr.gui.state.modeName}", switching to default`)
      await this.stateMgr.gui.updateState('modeName', DEFAULT_STATE.modeName)
    }
  }

  public async open () {
    await this._verifyWidgetStates()

    if (this.guiActive && this.panel) {
      this.panel.reveal()
      this.emit('webview.open')
      return
    }

    this.panel = vscode.window.createWebviewPanel(
      'marqueeGui', // Identifies the type of the webview. Used internally
      'Marquee', // Title of the this.panel displayed to the user
      vscode.ViewColumn.One, // Editor column to show the new webview panel in.
      {
        enableScripts: true,
        retainContextWhenHidden: true,
      }
    )
    this.panel.iconPath = {
      light: vscode.Uri.joinPath(this.context.extensionUri, 'assets', 'marquee-tab.svg'),
      dark: vscode.Uri.joinPath(this.context.extensionUri, 'assets', 'marquee-tab.svg'),
    }

    const basePath = vscode.Uri.joinPath(this._baseUri, 'dist', 'gui')
    const baseAppUri = this.panel.webview.asWebviewUri(basePath)
    const nonce = Buffer.from(uuidv4()).toString('base64')
    const config = vscode.workspace.getConfiguration('marquee')
    const pref: ExtensionConfiguration | undefined = config.get('configuration')
    // calculate font-size as range between 0.5 - 1.5
    const fontSize = 0.5 + ((typeof pref?.fontSize === 'number' ? pref.fontSize : DEFAULT_FONT_SIZE) / 10)
    const widgets = [
      /**
       * 3rd party VSCode extension widgets
       */
      ...vscode.extensions.all,
      /**
       * Marquee widgets
       */
      ...this.stateMgr.widgetExtensions
    ] as vscode.Extension<ExtensionExport>[]

    const widgetScripts: string[] = []
    for (const extension of widgets) {
      /**
       * continue if extension doesn't expose a marquee widget
       * within its package.json
       */
      if (
        !extension.packageJSON.marquee?.widget ||
        /**
         * don't show example widget extension in production
         */
        (process.env.NODE_ENV !== 'development' && extension.id === 'stateful.marquee')
      ) {
        continue
      }

      /**
       * only setup a communication channel to the extension backend, if
       */
      if (
      /**
         * the extension is active so we can access its exported APIs
         */
        extension.isActive &&
        /**
         * the extension properly exports a setup method
         */
        extension.exports &&
        extension.exports.marquee &&
        typeof extension.exports.marquee.setup === 'function'
      ) {
        const defaultState = extension.exports.marquee?.disposable?.state || {}
        const defaultConfiguration = extension.exports.marquee?.disposable?.configuration || {}
        const ch = new Channel(extension.id, { ...defaultState, ...defaultConfiguration })
        ch.registerPromise([this.panel.webview]).then((client) => {
          extension.exports.marquee.setup(client)
        })
      }

      /**
       * in order to allow accessing assets outside of the Marquee extension
       * we need to link to the directory as accessing files outside of the
       * extension dir is not possible (returns a 404)
       *
       * Only do this when extension host is running within Node.js environment.
       */
      if (extension.extensionPath && globalThis.process) {
        const extPath = path.join(this.context.extensionPath, THIRD_PARTY_EXTENSION_DIR, extension.id)
        const doesExist = await fs.access(extPath).then(() => true, () => false)
        if (!doesExist) {
          await fs.symlink(extension.extensionPath, extPath)
        }

        const targetPath = path.join(extPath, extension.packageJSON.marquee?.widget)
        const src = this.panel.webview.asWebviewUri(vscode.Uri.file(targetPath))
        widgetScripts.push(`<script type="module" src="${src.toString()}" nonce="${nonce}"></script>`)
      }
    }

    const aws = this.stateMgr.projectWidget.getActiveWorkspace()
    const cs = pref?.colorScheme
    const colorScheme = (
      typeof cs?.r === 'number' &&
      typeof cs?.g === 'number' &&
      typeof cs?.b === 'number' &&
      typeof cs?.a === 'number'
    )
      ? `rgba(${cs.r}, ${cs.g}, ${cs.b}, ${cs.a})`
      : 'transparent'

    const widgetStateConfigurations = this.stateMgr.widgetExtensions.reduce(
      (prev, curr) => ({
        ...prev,
        [curr.id]: {
          configuration: curr.exports.marquee?.disposable?.configuration || {},
          state: curr.exports.marquee?.disposable?.state || {},
        },
      }),
      {} as Record<string, Pick<ExtensionExport, 'configuration' | 'state'>>
    )

    const backendBaseUrl = vscode.Uri.parse(BACKEND_BASE_URL)
    const backendGeoUrl = vscode.Uri.parse(BACKEND_GEO_URL)
    const content = await render(await this.getTemplate(), {
      aws,
      nonce,
      fontSize,
      baseAppUri,
      colorScheme,
      props: getExtProps(),
      baseUrl: BACKEND_BASE_URL,
      geoUrl: BACKEND_GEO_URL,
      fwdGeoUrl: BACKEND_FWDGEO_URL,
      cspSource: this.panel.webview.cspSource,
      widgetStateConfigurations: Buffer.from(JSON.stringify(widgetStateConfigurations)).toString('base64'),
      widgetScripts,
      connectSrc: [
        `${backendBaseUrl.scheme}://${backendBaseUrl.authority}`,
        `${backendGeoUrl.scheme}://${backendGeoUrl.authority}`,
        'https://api.hackerwebapp.com',
        'https://*.ingest.sentry.io'
      ],
      childSrc: [
        'https://www.youtube.com',
        'https://player.vimeo.com',
        'https://fast.wistia.net'
      ]
    })

    if (!content) {
      return vscode.window.showErrorMessage('Couldn\'t load Marquee')
    }

    const ch = new Channel<MarqueeEvents>('vscode.marquee')
    this.client = await ch.registerPromise([this.panel.webview])
    this.client.on('telemetryEvent', ({ eventName, properties }) => (
      telemetry.sendTelemetryEvent(eventName, properties)))
    this.panel.webview.html = content
    this.panel.webview.onDidReceiveMessage(this._handleWebviewMessage.bind(this))
    this.panel.onDidDispose(this._disposePanel.bind(this))
  }

  private _disposePanel () {
    telemetry.sendTelemetryEvent('guiClose')
    this.guiActive = false
    this.panel = null
    this.emit('webview.close')
    this.client?.removeAllListeners()
    delete this.client
  }

  private async _handleWebviewMessage (e: any) {
    if (e.west && Array.isArray(e.west.execCommands)) {
      try {
        await Promise.all(e.west.execCommands.map(this._executeCommand.bind(this)))
        return
      } catch (err: any) {
        return vscode.window.showErrorMessage(`Marquee Error: ${(err as Error).message}`)
      }
    }

    if (e.west && e.west.notify && e.west.notify.message) {
      return this._handleNotifications(e.west.notify)
    }

    if (e.ready) {
      telemetry.sendTelemetryEvent('guiOpen')
      this.guiActive = true
      return this.emit('webview.open')
    }
  }

  private _executeCommand ({ command, args, options }: { command: string, args: any[], options: any }) {
    telemetry.sendTelemetryEvent('executeCommand', { command })
    if (args && args.length > 0 && command === 'vscode.openFolder') {
      return vscode.commands.executeCommand(command, vscode.Uri.parse(args[0].toString()), options)
    }

    if (args && args.length > 0) {
      return vscode.commands.executeCommand(command, args[0], options)
    }

    vscode.commands.executeCommand(command, undefined, options)
  }

  private _handleNotifications ({ type, message }: { type: string, message: string }) {
    switch (type) {
      case 'error':
        vscode.window.showErrorMessage(message)
        break
      case 'warning':
        vscode.window.showWarningMessage(message)
        break
      default:
        vscode.window.showInformationMessage(message)
    }
  }
}
