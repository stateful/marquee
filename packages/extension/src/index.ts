import type vscode from "vscode";
import type { Client } from 'tangle';

import telemetry from './telemetry';
import { MarqueeExtension } from "./extension";
import { getExtProps } from './utils';

export function activate(context: vscode.ExtensionContext) {
  telemetry.sendTelemetryEvent('extensionActivate', getExtProps());
  new MarqueeExtension(context);

  if (process.env.NODE_ENV !== 'development') {
    return {};
  }

  return {
    /**
     * demo code for example widget in "/example"
     */
    marquee: {
      setup: (tangle: Client<{ counter: number }>) => {
        let i = 0;
        setInterval(() => {
          tangle.emit('counter', ++i);
        }, 1000);
      }
    }
  };
}

// this method is called when your extension is deactivated
export function deactivate() {
  telemetry.sendTelemetryEvent('extensionDeactivate');
}
