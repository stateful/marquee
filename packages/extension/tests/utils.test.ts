import vscode from 'vscode';
import { isExpanded, filterByScope, activateGUI, linkMarquee } from '../src/utils';

jest.mock('@vscode-marquee/utils/extension', () => class {
  static defaultConfigurations = {
    'marquee.configuration.modes': { default: 'marquee.configuration.modes' },
    'marquee.configuration.proxy': { default: 'marquee.configuration.proxy' },
    'marquee.configuration.fontSize': { default: 'marquee.configuration.fontSize' },
    'marquee.configuration.launchOnStartup': { default: 'marquee.configuration.launchOnStartup' },
    'marquee.configuration.workspaceLaunch': { default: 'marquee.configuration.workspaceLaunch' }
  };
  setBroadcaster = jest.fn();
  broadcast = jest.fn();
  state = 'foobar';
  _disposables = [];
  _channel = {
    appendLine: jest.fn()
  };
  configuration = {} as any;
  updateConfiguration (prop: string, val: any) {
    this.configuration[prop] = val;
  }
});

jest.mock('../src/constants', () => ({ MODES_UPDATE_TIMEOUT: 100 }));

test('isExpanded', () => {
  expect(isExpanded(1)).toBe('foo');
  expect(isExpanded(4)).toBe('bar');
  expect(isExpanded(-2)).toBe('bar');
});

test('filterByScope', () => {
  const obj = { workspaceId: '123' };
  expect(filterByScope([obj], null, true)).toEqual([obj]);
  expect(filterByScope([obj], null, false)).toEqual([]);
  expect(filterByScope([obj], { id: '321' }, false)).toEqual([]);
  expect(filterByScope([obj], { id: '123' }, false)).toEqual([obj]);
});

test('activateGUI', () => {
  const context = {
    globalState: {
      get: jest.fn().mockReturnValue({}),
      setKeysForSync: jest.fn()
    }
  };
  const extExport = activateGUI(context as any, {} as any);
  expect(extExport.marquee.disposable.state).toEqual('foobar');
});

test('extension manager removes native icon', () => {
  const context = {
    globalState: {
      get: jest.fn().mockReturnValue({}),
      setKeysForSync: jest.fn()
    }
  };
  const extExport = activateGUI(context as any, {} as any);
  extExport.marquee.disposable.updateConfiguration('modes', {
    foobar: {
      icon: { foo: 'bar', native: 123 }
    }
  });
  expect(extExport.marquee.disposable.configuration).toEqual({
    modes: {
      foobar: { icon: { foo: 'bar' } }
    }
  });
});

test('extension manager listens on mode changes and applies if not triggered within the last 2s', async () => {
  (vscode.workspace.getConfiguration as jest.Mock).mockReturnValueOnce({
    get: jest.fn().mockReturnValue({ foo: 'bar' })
  });
  const context = {
    globalState: {
      get: jest.fn().mockReturnValue({}),
      setKeysForSync: jest.fn()
    }
  };
  const extExport = activateGUI(context as any, {} as any);
  expect(extExport.marquee.disposable['broadcast']).toBeCalledTimes(0);

  extExport.marquee.disposable['_onModeChange']({ affectsConfiguration: jest.fn().mockReturnValue(false) });
  expect(extExport.marquee.disposable['broadcast']).toBeCalledTimes(0);

  extExport.marquee.disposable['_onModeChange']({ affectsConfiguration: jest.fn().mockReturnValue(true) });
  expect(extExport.marquee.disposable['broadcast']).toBeCalledTimes(0);

  await new Promise((resolve) => setTimeout(resolve, 110));

  extExport.marquee.disposable['_onModeChange']({ affectsConfiguration: jest.fn().mockReturnValue(true) });
  expect(extExport.marquee.disposable['broadcast']).toBeCalledTimes(1);
  expect(extExport.marquee.disposable['broadcast']).toBeCalledWith({
    modes: { foo: 'bar' }
  });
});

test('linkMarquee', async () => {
  const parse = vscode.Uri.parse as jest.Mock;
  await linkMarquee({ item: { origin: '/some/file:123:3' } });
  expect(parse).toBeCalledWith('/some/file:123');
  await linkMarquee({ item: { origin: '/some/file:124' } });
  expect(parse).toBeCalledWith('/some/file');
});
