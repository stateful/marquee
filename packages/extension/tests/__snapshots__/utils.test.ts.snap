// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should get proper extension props 1`] = `
Object {
  "extname": "stateful.marquee",
  "extversion": "3.0.0-edge.1",
  "os": "some platform",
  "platformversion": "some release",
  "uikind": "desktop",
  "vscodemachineid": "machineId",
  "vscodesessionid": "sessionId",
  "vscodeversion": "1.2.3",
}
`;
